Update MediaSource.cpp to comply with new CPP-safer rules.

jyavenard opened the pull request:
#### 02f76ccb0345a51164a5f1e7a57a774cf7e445eb
<pre>
Update MediaSource.cpp to comply with new CPP-safer rules.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282303">https://bugs.webkit.org/show_bug.cgi?id=282303</a>
<a href="https://rdar.apple.com/138878642">rdar://138878642</a>

Reviewed by Youenn Fablet.

Make SourceBufferList use Vector&lt;Ref&lt;SourceBuffer&gt;&gt; as storage.
Adopt more smart pointers to access SourceBuffer and MediaSourcePrivate.

No observable change.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::setStreaming):
(WebCore::ManagedMediaSource::streamingTimerFired):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::reOpen):
(WebCore::MediaSource::duration const):
(WebCore::MediaSource::currentTime const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::waitForTarget):
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::setLiveSeekableRange):
(WebCore::MediaSource::clearLiveSeekableRange):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::isBuffered const):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDuration):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::setReadyState):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::openIfInEndedState):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::createSourceBufferPrivate):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::sourceBufferReceivedFirstInitializationSegmentChanged):
(WebCore::MediaSource::sourceBufferActiveTrackFlagChanged):
(WebCore::MediaSource::setMediaPlayerReadyState):
(WebCore::MediaSource::protectedPrivate const):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBufferList.cpp:
(WebCore::SourceBufferList::contains const):
(WebCore::SourceBufferList::item const):
(WebCore::SourceBufferList::remove):
(WebCore::SourceBufferList::replaceWith):
* Source/WebCore/Modules/mediasource/SourceBufferList.h:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/285899@main">https://commits.webkit.org/285899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d787c97c4d0dfd87d64d7ded8c0d114e43ed2cc1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26914 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78492 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25341 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58271 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16602 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77170 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63740 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38681 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21236 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23674 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66791 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79995 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/776 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66561 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1564 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65837 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16328 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9762 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4172 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1413 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -77,8 +77,8 @@ void ManagedMediaSource::setStreaming(bool streaming)
     m_streaming = streaming;
-    if (m_private)
-        m_private->setStreaming(streaming);
+    if (RefPtr msp = protectedPrivate())
```
RefPtr msp = m_private works too.

jyavenard replied:
yes, but I figured it looked nicer that way

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -361,20 +361,24 @@ MediaTime MediaSource::duration() const
-    return m_private ? m_private->duration() : MediaTime::invalidTime();
+    if (RefPtr msp = protectedPrivate())
+        return msp->duration();
+    return MediaTime::invalidTime();
```
Isn't duration a simple getter?

jyavenard replied:
it's not inline, so it won't be considered trivial

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -361,20 +361,24 @@ MediaTime MediaSource::duration() const
 
-    return m_private ? m_private->currentTime() : MediaTime::zeroTime();
+    if (RefPtr msp = protectedPrivate())
+        return msp->currentTime();
```
Ditto

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -463,14 +471,16 @@ void MediaSource::completeSeek()
 {
-    return m_private ? TimeRanges::create(m_private->seekable()) : TimeRanges::create();
+    if (RefPtr msp = protectedPrivate())
+        return TimeRanges::create(msp->seekable());
```
Ditto.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -384,6 +388,10 @@ Ref<MediaTimePromise> MediaSource::waitForTarget(const SeekTarget& target)
 
+    RefPtr msp = protectedPrivate();
+    if (!msp)
+        return MediaTimePromise::createAndReject(PlatformMediaError::SourceRemoved);
```
This is new, is this needed, or should we use releaseNonNull()?

jyavenard replied:
it is new, and while I don't think it could happen, I wanted to cover the case.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -505,7 +517,8 @@ ExceptionOr<void> MediaSource::clearLiveSeekableRange()
         return Exception { ExceptionCode::InvalidStateError };
-    m_private->clearLiveSeekableRange();
+    Ref msp = protectedPrivate().releaseNonNull();
+    msp->clearLiveSeekableRange();
```
Why not simply protectedPrivate()->clearLiveSeekableRange();

jyavenard replied:
I believe this provides a clearer way to indicate that the RefPtr returned isn't null (as it's called after the check `(isClosed())`

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -549,7 +563,9 @@ bool MediaSource::hasFutureTime()
         return false;
 
-    return m_private->hasFutureTime(currentTime(), m_private->timeIsProgressing() ? MediaTime::zeroTime() : MediaSourcePrivate::futureDataThreshold());
+    Ref msp = protectedPrivate().releaseNonNull();
```
Ditto

jyavenard replied:
same. it's very explicit here that msp is a Ref and is non-null.

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1702,6 +1730,11 @@ Ref<SourceBufferList> MediaSource::activeSourceBuffers() const
     return m_activeSourceBuffers;
 }
 
+RefPtr<MediaSourcePrivate> MediaSource::protectedPrivate() const
```
Should we inline it?

jyavenard replied:
it gives compilation error as MediaSourcePrivate is only fast forwarded; I would need to pull the entire MediaSourcePrivate.h header

youennf commented on Source/WebCore/Modules/mediasource/SourceBufferList.cpp:
```diff
@@ -67,9 +67,21 @@ void SourceBufferList::add(Ref<SourceBuffer>&& buffer)
 void SourceBufferList::remove(SourceBuffer& buffer)
 {
-    size_t index = m_list.find(&buffer);
+    size_t index = m_list.find(Ref { buffer });
```
find(buffer) is not working?

jyavenard replied:
unfortunately not. 
the definition and templating definition of find gave me an error

jyavenard committed:
Update MediaSource.cpp to comply with new CPP-safer rules.
https://bugs.webkit.org/show_bug.cgi?id=282303
rdar://138878642

Reviewed by Youenn Fablet.

Make SourceBufferList use Vector<Ref<SourceBuffer>> as storage.
Adopt more smart pointers to access SourceBuffer and MediaSourcePrivate.

No observable change.

* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
(WebCore::ManagedMediaSource::setStreaming):
(WebCore::ManagedMediaSource::streamingTimerFired):
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setPrivateAndOpen):
(WebCore::MediaSource::reOpen):
(WebCore::MediaSource::duration const):
(WebCore::MediaSource::currentTime const):
(WebCore::MediaSource::buffered const):
(WebCore::MediaSource::waitForTarget):
(WebCore::MediaSource::seekToTime):
(WebCore::MediaSource::seekable):
(WebCore::MediaSource::setLiveSeekableRange):
(WebCore::MediaSource::clearLiveSeekableRange):
(WebCore::MediaSource::hasBufferedTime):
(WebCore::MediaSource::hasFutureTime):
(WebCore::MediaSource::isBuffered const):
(WebCore::MediaSource::monitorSourceBuffers):
(WebCore::MediaSource::setDuration):
(WebCore::MediaSource::setDurationInternal):
(WebCore::MediaSource::setReadyState):
(WebCore::MediaSource::streamEndedWithError):
(WebCore::MediaSource::detachFromElement):
(WebCore::MediaSource::openIfInEndedState):
(WebCore::MediaSource::readyState const):
(WebCore::MediaSource::createSourceBufferPrivate):
(WebCore::MediaSource::regenerateActiveSourceBuffers):
(WebCore::MediaSource::updateBufferedIfNeeded):
(WebCore::MediaSource::sourceBufferReceivedFirstInitializationSegmentChanged):
(WebCore::MediaSource::sourceBufferActiveTrackFlagChanged):
(WebCore::MediaSource::setMediaPlayerReadyState):
(WebCore::MediaSource::protectedPrivate const):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/Modules/mediasource/SourceBufferList.cpp:
(WebCore::SourceBufferList::contains const):
(WebCore::SourceBufferList::item const):
(WebCore::SourceBufferList::remove):
(WebCore::SourceBufferList::replaceWith):
* Source/WebCore/Modules/mediasource/SourceBufferList.h:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:

Canonical link: https://commits.webkit.org/285899@main
